src/Servers/Kestrel/Transport.MsQuic/Internal/QuicSession.cs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using static Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal.QuicListener;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Transport.MsQuic.Internal
+{
+    // TODO I still need to finish a small refactor here to remove allocations per connection.
+    internal sealed class QuicSession : IDisposable
+    {
+        private bool _disposed = false;
+        private IntPtr _nativeObjPtr;
+        private byte[] _alpn;
+        private QuicRegistration _registration;
+
+        internal QuicSession(QuicRegistration registration, IntPtr nativeObjPtr, byte[] alpn)
+        {
+            _registration = registration;
+            _alpn = alpn;
+            _nativeObjPtr = nativeObjPtr;
+        }
+
+        public QuicListener ListenerOpen(ListenerCallback callback)
+        {
+            var listenerPtr = IntPtr.Zero;
+            var status = (QUIC_STATUS)_registration.ListenerOpenDelegate(
+                _nativeObjPtr,
+                NativeCallbackHandler,
+                IntPtr.Zero,
+                out listenerPtr
+                );
+
+            QuicStatusException.ThrowIfFailed(status);
+            var listener = new QuicListener(
+                _registration,
+                listenerPtr,
+                true);
+
+            listener.SetCallbackHandler(callback);
+            return listener;
+        }
+
+        public ValueTask<QuicConnection> ConnectionOpenAsync(
```

Comment:
This function isn't actually async.

Reply:
Yeah this should be async. It should wait for the connected event to occur. I haven't invested a ton into the client side.